Add built-in math functions to use in lambdas as in NCalc

eugenca opened the pull request:
Only changes in LambdaExpressionVistor.cs

eugenca commented:
You can check this is actually working in lambda here: https://eugenca.github.io/wasm_projects/runge-kutta/

eugenca commented:
Actually, 12 000 000 evaluations (4 per step of 3kk steps) of ```sin(x)+cos(sin(x))-tan(log10(x+2))``` plus some math (rk4) in 2908ms in browser wasm.
C++ could make 72 000 000 evaluations (in absolutely cache-friendly loop)

david-brink-talogy commented on src/NCalc/LambdaExpressionVistor.cs:
```diff
@@ -200,26 +230,76 @@ public override void Visit(Function function)
+            {
+                CheckArgumentsLengthForFunction(mathCallFuncName, argsNumActual, 1);
+                var arg = L.Expression.Convert(args[0], typeof(double));
+                _result = L.Expression.Call(typeof(Math).GetMethod(mathCallFuncName), arg);
```
It seems like this would use reflection much more than necessary.  Every time a math function is encountered you're using reflection to get the method.  I think you should be able to do the lookup once via a static property and reuse it indefinitely.

eugenca replied:
Sounds good

david-brink-talogy commented on src/NCalc/LambdaExpressionVistor.cs:
```diff
@@ -200,26 +230,76 @@ public override void Visit(Function function)
             }
 
-            switch (functionName)
+            Action<string, int, int> CheckArgumentsLengthForFunction = delegate (string funcStr, int argsNum, int argsNeed)
```
Does this need to be a delegate or can it be a normal function call?

eugenca commented:
@david-brink-talogy @sklose  this is experimental behavior, and it could break some existing code, so I ask about possibilities of existence of that functionality

eugenca commented on src/NCalc/LambdaExpressionVistor.cs:
```diff
@@ -200,26 +230,76 @@ public override void Visit(Function function)
+                _result = L.Expression.Call(typeof(Math).GetMethod(mathCallFuncName), arg);
+            };
+
+            Action<string, int> MakeMathCallExpressionTwoDoubleParam = delegate (string mathCallFuncName, int argsNumActual)
```
I wanted anonymous function inside method with small reusable functionality

david-brink-talogy replied:
I think what you want is a [local function](https://learn.microsoft.com/en-us/dotnet/csharp/programming-guide/classes-and-structs/local-functions), though I doubt you can use them here due to c# version requirements.  

There's overhead to declaring delegates like this.  Every time Visit is called you're declaring these delegates as a local variable.  A private static class method would perform much better. 

I wrote up a quick test to illustrate.

![image](https://github.com/sklose/NCalc2/assets/43828739/fad44ed7-459f-41d7-853f-901bda5f560f)
```
BenchmarkRunner.Run<TestInvoke>();

public class TestInvoke
{
    [Benchmark(Baseline = true)]
    public void InvokeDelegate()
    { 
        Action<string, int> MakeMathCallExpressionTwoDoubleParam = delegate (string mathCallFuncName, int argsNumActual)
        {
            if (argsNumActual > 1000)
            {
                throw new ArgumentOutOfRangeException();
            }
        };

        MakeMathCallExpressionTwoDoubleParam("one", 1);
    }

    [Benchmark]
    public void InvokeMethod()
    {
        MakeMathCallExpressionTwoDoubleParam("one", 1);
    }

    private static void MakeMathCallExpressionTwoDoubleParam(string mathCallFuncName, int argsNumActual)
    {
        if (argsNumActual > 1000)
        {
            throw new ArgumentOutOfRangeException();
        }
    }
}

```

eugenca replied:
Thanks! Local methods are from C#7 so according to this
https://learn.microsoft.com/en-us/dotnet/csharp/language-reference/configure-language-version
It may have support

eugenca replied:
But I found that .NET Framework 4.5 use C# 5

eugenca replied:
If I could successfully create nupkg does it mean that I could use C# 7?

david-brink-talogy replied:
I don't know the answer.  The compiler could simply remove all traces of the local function via inlining.

eugenca commented:
> Can you please attach BenchmarkDotNet tests to compare performance with the current version?

Performance is just the same.

Sample code (I made internal dictionary to have SINX and COSX to test performance)
```
public struct Context
{
    public double x;
    public double Sin(double x) { return Math.Sin(x); }
    public double Cos(double x) { return Math.Cos(x); }
}

funcInternal = FunctionParser.Parse<Context>("sinx(x) + cosx(x)");
funcContextMethods = FunctionParser.Parse<Context>("sin(x) + cos(x)");
context = new Context() { x = 5 };
```

Results:
```
| Method          | Mean     | Error    | StdDev   |
|---------------- |---------:|---------:|---------:|
| RunInternalFunc | 10.49 ns | 0.175 ns | 0.164 ns |
| RunContextFunc  | 10.66 ns | 0.021 ns | 0.020 ns |
```

sklose commented:
Hi @eugenca,

I like adding support for these functions to the lambda compilation visitor. I didn't get a chance to review these changes in detail. 2 things I would like to ensure with this:

1. the interpreter and the compiler have support for the same set of functions with the same behavior
2. the user can still override these functions as part of the context (that way it's not a breaking change and the behavior is customizable)

Happy to get this merged if we can ensure these two things.

eugenca commented:
Ok! I will try to ensure this. If anyone have additional ideas for tests please share

eugenca commented:
@sklose 
Added tests to ensure your requirements. Tested Lambda with double types only for now

@david-brink-talogy 
Added method caching into static dictionary
But it seems that getting method without caching brings no extra overhead compared to lambda creation. I felt no difference with or without cache

david-brink-talogy commented on src/NCalc/LambdaExpressionVistor.cs:
```diff
@@ -26,6 +26,37 @@ internal class LambdaExpressionVistor : LogicalExpressionVisitor
+            public int ArgumentCount;
+        }
+
+        private readonly Dictionary<string, MathCallFunction> _mathCallFunctions = new Dictionary<string, MathCallFunction>()
```
@eugenca, In regards to only using reflection once, here's what I was thinking.  _mathCallFunctions can be static so that it is only instantiated and initialized once.  The initialization can also look up the method info, since you're only ever calling the overload that takes double(s). 

```
private struct MathCallFunction
{
    public MethodInfo MethodInfo;
    public int ArgumentCount;
}

private static readonly Dictionary<string, MathCallFunction> _mathCallFunctions = new() 
{
  { "ABS", new MatchCallFunction(typeof(Math).GetMethod(nameof(Math.Abs), new[typeof(double)]), 1) },
  //...
  { "ATAN2", new MatchCallFunction(typeof(Math).GetMethod(nameof(Math.Atan2), new[typeof(double), typeof(double)]), 2) },
  //...
};
```

eugenca replied:
Looks reasonable

sklose commented:
LGTM, just needs to be rebased

eugenca commented:
Maybe it is worth to update readme with examples or links, to allow users to know about this functionality
